packages/query/src/block-processor.ts

```diff
@@ -169,171 +187,178 @@ export class BlockProcessor implements BlockProcessorInterface {
+     *
+     * You can see above that we're already iterating over flush.newNotes. So
+     * why don't we put this call to
+     * `this.maybeUpsertAuctionWithNoteCommitment()` inside that earlier `for`
+     * loop?
+     *
+     * The problem is, we need to call `this.processTransactions()` before
+     * calling `this.maybeUpsertAuctionWithNoteCommitment()`, because
+     * `this.processTransactions()` is what saves the auction NFT metadata to
+     * the database. `this.maybeUpsertAuctionWithNoteCommitment()` depends on
+     * that auction NFT metadata being saved already to be able to detect
+     * whether a given note is for an auction NFT; only then will it save the
+     * note's commitment to the `AUCTIONS` table.
+     *
+     * "So why not just move `this.processTransactions()` to before the block
+     * where we handle `flush.newNotes`?" Because `this.processTransactions()`
+     * should only run after we've handled `flush.newNotes`, since we depend
+     * on the result of the flush to determine whether there are transactions
+     * to process in the first place. It's a catch-22.
+     *
+     * This isn't a problem in core because core isn't going back and forth
+     * between Rust and TypeScript like we are. If and when we move the block
+     * processor into Rust, this issue should be resolved.
+     */
+    for (const spendableNoteRecord of flush?.newNotes ?? []) {
+      await this.maybeUpsertAuctionWithNoteCommitment(spendableNoteRecord);
+    }
 
-      const isLastBlockOfEpoch = !!compactBlock.epochRoot;
-      if (isLastBlockOfEpoch) {
-        await this.handleEpochTransition(compactBlock.height, latestKnownBlockHeight);
-      }
+    // We do not store historical prices,
+    // so there is no point in saving prices that would already be considered obsolete at the time of saving
+    const blockInPriceRelevanceThreshold =
+      compactBlock.height >= latestKnownBlockHeight - BigInt(PRICE_RELEVANCE_THRESHOLDS.default);
 
-      if (globalThis.__ASSERT_ROOT__) {
-        await this.assertRootValid(compactBlock.height);
-      }
+    // we can't use third-party price oracles for privacy reasons,
+    // so we have to get asset prices from swap results during block scans
+    // and store them locally in indexed-db.
+    if (blockInPriceRelevanceThreshold && compactBlock.swapOutputs.length) {
+      await updatePricesFromSwaps(
+        this.indexedDb,
+        this.numeraires,
+        compactBlock.swapOutputs,
+        compactBlock.height,
+      );
+    }
+
+    // We only query Tendermint for the latest known block height once, when
+    // the block processor starts running. Once we're caught up, though, the
+    // chain will of course continue adding blocks, and we'll keep processing
+    // them. So, we need to update `latestKnownBlockHeight` once we've passed
+    // it.
+    if (compactBlock.height > latestKnownBlockHeight) {
+      latestKnownBlockHeight = compactBlock.height;
```

Comment:
This should move to the loop within `syncAndStore()` so we don't need to override the input

Reply:
we'll have to move 

```
      const isLastBlockOfEpoch = !!compactBlock.epochRoot;
      if (isLastBlockOfEpoch) {
        await this.handleEpochTransition(compactBlock.height, latestKnownBlockHeight);
      }
```

since handling the epoch transition also relies on `latestKnownBlockHeight`